garrysmod/lua/includes/extensions/player.lua

```diff
@@ -87,6 +87,108 @@ function player.GetBySteamID64( ID )
 	return false
 end
 
+if (SERVER) then
+	util.AddNetworkString("GMOD_player_SendMsg_Text")
+	util.AddNetworkString("GMOD_player_SendMsg_Color")
+
+	local function SendMsgError(uArg, v)
+		error(string.format("bad argument #%u to player.SendMsg (string or Color expected, got %s)",
+			uArg + 1, type(v)), 3)
+	end
+
+	function player.SendMsg(filter, ...)
+		local tParts = {...}
+		local bLastString = false
+
+		local bFirstColor = false
+		local tWriteParts = {}
+		local uWriteParts = 0
+
+		for i = 1, select("#", ...) do
+			local part = tParts[i]
+
+			if (bLastString) then
+				if (IsTableColor(part, false)) then

```

Comment:
This is where it's wrong, stick to a format, don't "guess"/detect if it's STRING->COLOR->STRING->COLOR etc or STRING->STRING->COLOR->STRING

Reply:
It's not guessing. Instead of networking STRING->STRING->COLOR->STRING, the strings are concated so it sticks to a consistent format of alternating strings and colors. Aka what the engine already does with chat.AddText.

Reply:
Yes my bad on this, I mixed up the client and the server.
Anyway from the start the function should not be called this way.

`assert` table and assert string, fuck him if he doesn't call it correctly

Reply:
There should be no "fuck him," there should just be proper type checking like what's done here. You can send a string first or a color first. What's the actual issue here?

Reply:
I think this is where we disagree, I don't think it's a good idea to allow multiple way to call a function.

Reply:
Well this is how chat.AddText behaves. It's too late to change it, and this is not just some garry-ism we can just ignore as it is the only way to add coloured text (or add any text clientside).

Reply:
Sure it's too late to change chat.AddText or to remove it, but I don't think we can justify everything by "that's how it is".

for example, table.HasValue exists, yet, thing kind of stuff should not be "repeated".

Reply:
There is an easy alternative to using table.HasValue - you can even structure your tables around it. There is no alternative chat.AddTexf alternative and it's fine to maintain this consistency.